kafka-client/src/main/java/dev/responsive/kafka/internal/db/WindowedKeySpec.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2023 Responsive Computing, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.db;
+
+import static org.apache.kafka.streams.state.StateSerdes.TIMESTAMP_SIZE;
+
+import dev.responsive.kafka.internal.utils.WindowedKey;
+import java.nio.ByteBuffer;
+import java.util.function.Predicate;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.utils.Bytes;
+
+public class WindowedKeySpec implements KeySpec<WindowedKey> {
+
+  private final Predicate<WindowedKey> withinRetention;
+
+  public WindowedKeySpec(final Predicate<WindowedKey> withinRetention) {
+    this.withinRetention = withinRetention;
+  }
+
+  @Override
+  public int sizeInBytes(final WindowedKey key) {
+    return key.key.get().length + Long.BYTES;
+  }
+
+  @Override
+  public WindowedKey keyFromRecord(final ConsumerRecord<byte[], byte[]> record) {
+    final byte[] key = record.key();
+
+    // the WindowKeySchema in Streams always encodes the changelog record keys by
```

Comment:
since this has bitten us before, what do you think about using `WindowKeySchema#fromStoreBytesKey` directly and then just discarding things we don't use (like the window size)? 

It also looks like it's encoded different based on whether its from a changelog topic or a source topic?
```
        // toStoreKeyBinary was used to serialize the data.
        if (this.isChangelogTopic) {
            return WindowKeySchema.fromStoreKey(data, windowSize, inner, topic);
        }

        // toBinary was used to serialize the data
        return WindowKeySchema.from(data, windowSize, inner, topic);
```

`seqNum` is ignored in the latter, which is a bit confusing?

We should make sure to have a test where we source directly from a windowed output topic w/o a changelog

Reply:
I'll take a look at the WindowKeySchema to make sure I'm not missing anything, but we definitely do not support windowed source tables or WindowStores with a source-topic changelog. See [KIP-300](https://cwiki.apache.org/confluence/display/KAFKA/KIP-300%3A+Add+Windowed+KTable+API+in+StreamsBuilder) for context (note that this KIP was abandoned so the changes proposed there do not exist)

Reply:
Oh wait, unfortunately the WindowKeySchema only converts bytes to a full `Windowed<Bytes>` key, which means it does an extra computation to compute the end time. It wouldn't be the end of the world to just pass in a dummy window size and throw out the end time, but it seems silly to do this extra work when we can just extract only the needed bits. I know it's just a bit of addition but this is very much on the hot path, so may as well cut it out

However, I will add a unit test that verifies the WindowKeySchema format encodes and decodes the way we expect it to, just to be safe